File: Laravel/app/Biab/Resources/ParallelPages.php
<?php

namespace App\Biab\Resources;

use App\Biab\Client;

/**
 * Programmatic SEO. One template × N variant tuples — `/services/{service}/{area}`
 * and friends — rendered by the platform so the copy stays editable in the
 * dashboard rather than hard-coded here.
 *
 * `listVariants()` is what a sitemap iterates.
 */
class ParallelPages
{
    public function __construct(private readonly Client $client)
    {
    }

    /** @return array<string, mixed> */
    public function list(): array
    {
        return $this->client->get($this->client->sitePath('parallel-pages'));
    }

    /** @return array<string, mixed> */
    public function listVariants(string $key): array
    {
        return $this->client->get(
            $this->client->sitePath('parallel-pages/'.rawurlencode($key).'/variants')
        );
    }

    /**
     * @param  array<string, string>  $params  one value per variant axis
     * @return array<string, mixed>
     */
    public function render(string $key, array $params): array
    {
        return $this->client->get(
            $this->client->sitePath('parallel-pages/'.rawurlencode($key).'/render'),
            $params
        );
    }
}

File: Phoenix/lib/bd_starter/bd/resources/parallel_pages.ex
defmodule BdStarter.Bd.Resources.ParallelPages do
  @moduledoc """
  Programmatic SEO. One template × N variant tuples, rendered by the platform
  so a hundred landing pages stay editable in the dashboard instead of being
  generated and forgotten in this repo.

  `variants/2` is what a sitemap iterates.
  """

  alias BdStarter.Bd.Client

  def list(client), do: Client.get(client, Client.site_path(client, "parallel-pages"))

  def variants(client, key),
    do: Client.get(client, Client.site_path(client, "parallel-pages/#{enc(key)}/variants"))

  def render(client, key, params),
    do: Client.get(client, Client.site_path(client, "parallel-pages/#{enc(key)}/render"), Map.to_list(params))

  defp enc(v), do: URI.encode(to_string(v), &URI.char_unreserved?/1)
end
